### PR TITLE
LockFiles with adjustments

### PR DESCRIPTION
**Pull request by pjf**
This is #1265, but with the addition of tests and extra garbage collection. Specifically we need reviewed:

* a43ec69 (garbage collection)
* b60da6c (tests)

### CODE REVIEW

**Comment by Postremus** [comment]
I opened the gui. I then wanted to install OPM with the cli. Got this error:

```
C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Cmdline\bin\Debug>Cmd
Line.exe install OuterPlanetsMod

Unbehandelte Ausnahme: CKAN.RegistryInUseKraken: Eine Ausnahme vom Typ "CKAN.Reg
istryInUseKraken" wurde ausgelöst.
   bei CKAN.RegistryManager..ctor(String path, KSP ksp) in C:\Users\Martin\Docum
ents\Visual Studio 2015\Projects\CKAN\Core\Registry\RegistryManager.cs:Zeile 49.

   bei CKAN.RegistryManager.Instance(KSP ksp) in C:\Users\Martin\Documents\Visua
l Studio 2015\Projects\CKAN\Core\Registry\RegistryManager.cs:Zeile 146.
   bei CKAN.KSP.get_RegistryManager() in C:\Users\Martin\Documents\Visual Studio
 2015\Projects\CKAN\Core\KSP.cs:Zeile 33.
   bei CKAN.KSP.get_Registry() in C:\Users\Martin\Documents\Visual Studio 2015\P
rojects\CKAN\Core\KSP.cs:Zeile 38.
   bei CKAN.KSP.ScanGameData() in C:\Users\Martin\Documents\Visual Studio 2015\P
rojects\CKAN\Core\KSP.cs:Zeile 321.
   bei CKAN.CmdLine.MainClass.Scan(KSP ksp_instance, IUser user, String next_com
mand) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Cmdline\Main
.cs:Zeile 338.
   bei CKAN.CmdLine.MainClass.Main(String[] args) in C:\Users\Martin\Documents\V
isual Studio 2015\Projects\CKAN\Cmdline\Main.cs:Zeile 188.
```

**Comment by pjf** [comment]
> I opened the gui. I then wanted to install OPM with the cli. Got this error:

Hooray! We should probably make that more user-friendly, shouldn't we?

But hooray for the correct kraken being thrown.

**Comment by Postremus** [comment]
Yes, the user friendly message should be shown.

I ran the Tests. Most of the throw a "The process cannot access the file "registry.json.locked" because it is being used by another process." now.

File: Core/Registry/RegistryManager.cs [hunk 1]
```diff
@@ -49,6 +61,78 @@ private RegistryManager(string path, KSP ksp)
...
+        /// </summary>
+        /// <returns><c>true</c>, if lock was gotten, <c>false</c> otherwise.</returns>
+        public bool GetLock()
+        {
+            try
+            {
+                lockfile_stream = new FileStream(lockfile_path, FileMode.CreateNew, FileAccess.Write, FileShare.None, 512, FileOptions.DeleteOnClose);
+
+                // Write the current process ID to the file.
+                StreamWriter writer = new StreamWriter(lockfile_stream);
```
Comment (Postremus) [inline Core/Registry/RegistryManager.cs]: The StreamWriter should be closed and disposed.

Comment (pjf) [inline Core/Registry/RegistryManager.cs]: Oh, good catch! We need to keep the `FileStream` open, but not the `StreamWriter`.

Comment (pjf) [inline Core/Registry/RegistryManager.cs]: But we can't *close* the StreamWriter, as that would close the underlying file.

Comment (pjf) [inline Core/Registry/RegistryManager.cs]: And we can't dispose it either, as that also releases the lock. Adjusting code now.

**Comment by pjf** [comment]
Updates applied include:

* Refactoring of `CmdLine/Main` to simplify the switch statement and code flow.
* Fixed an instance of trying to clean up after discovering a lock resulting in... another lock error.
* Made the error message a little more user friendly ("CKAN already running" rather than "registry locked")
* StreamWriter is disposed of on lock release.

**Comment by Postremus** [comment]
Opening the gui, then closing it throws this:

```
Unhandled exception:
System.ObjectDisposedException: Auf eine geschlossene Datei kann nicht zugegriffen werden.
   bei System.IO.__Error.FileNotOpen()
   bei System.IO.FileStream.Flush(Boolean flushToDisk)
   bei System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   bei System.IO.StreamWriter.Dispose(Boolean disposing)
   bei System.IO.TextWriter.Dispose()
   bei CKAN.RegistryManager.ReleaseLock() in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Core\Registry\RegistryManager.cs:Zeile 141.
   bei CKAN.RegistryManager.Dispose(Boolean safeToAlsoFreeManagedObjects) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Core\Registry\RegistryManager.cs:Zeile 89.
   bei CKAN.RegistryManager.Dispose() in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Core\Registry\RegistryManager.cs:Zeile 79.
   bei CKAN.Main..ctor(String[] cmdlineArgs, GUIUser User, Boolean showConsole) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Main.cs:Zeile 240.
   bei CKAN.GUI.Main_(String[] args, Boolean showConsole) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Program.cs:Zeile 35.
   bei CKAN.GUI.Main(String[] args) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Program.cs:Zeile 17.
   bei System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   bei System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   bei Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   bei System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   bei System.Threading.ThreadHelper.ThreadStart() 
```

**Comment by pjf** [comment]
@Postremus : Resource disposal adjusted! Travis seems to like this better, too. :)

**Comment by Postremus** [comment]
The Tests still throw an "The process cannot access the file "registry.json.locked" because it is being used by another process." at me.

**Comment by pjf** [comment]
@Postremus : I've found a very obvious suspect for the locking issues when running tests! I hope badfc43 brings great joy!

**Comment by pjf** [comment]
@Postremus : And another patch, but I suspect you'll see at least one more bug, but we're *almost* there. :)

File: GUI/Main.cs [hunk 2]
```diff
@@ -187,6 +187,21 @@ public Main(string[] cmdlineArgs, GUIUser User, bool showConsole)
                     Repo.default_ckan_repo.ToString()
                 );
 
+            // Check if there is any other instances already running.
+            // This is not entirely necessary, but we can show a nicer error message this way.
+            try
+            {
+                #pragma warning disable 219
```
Comment (Postremus) [inline GUI/Main.cs]: Why is this needed?

File: Core/Types/Kraken.cs [hunk 3]
```diff
@@ -299,4 +299,20 @@ public override string ToString()
...
+
+        public RegistryInUseKraken(string lockfile_path, string reason = null, Exception inner_exception = null)
+            :base(reason, inner_exception)
+        {
+            this.lockfile_path = lockfile_path;
+        }
+
+        public override string ToString()
+        {
+            return String.Format("CKAN is already running for this instance!\n\nIf you're certain this is not the case, then delete:\n\"{0}\"\n", lockfile_path);
```
Comment (Postremus) [inline Core/Types/Kraken.cs]: Maybe use "Enviroment.NewLine" for the new line?

File: GUI/Main.cs (hunk 2 above)
Comment (pjf) [inline GUI/Main.cs]: This may be one for @mgsdk (who authored the change). I suspect calling `RegistryManager.Instance(CurrentInstance).registry` without an assignment generates a warning and/or requires an assignment, but I honestly haven't tested that.

Comment (mgsdk) [inline GUI/Main.cs]: It is for assigning a variable but never using it, [CS0219](https://msdn.microsoft.com/en-us/library/4kycwe2x(v=vs.90).aspx). Since I know it wouldn't be used, I thought I would reduce the warnings in the output.

**Comment by Postremus** [comment]
1) Needs a rebase
2) Got this exception when selecting an instance.
```
Unhandled exception:
System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei CKAN.KSPManager.SetCurrentInstance(String name) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\Core\KSPManager.cs:Zeile 238.
   bei CKAN.ChooseKSPInstance.UseSelectedInstance() in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\ChooseKSPInstance.cs:Zeile 86.
   bei CKAN.ChooseKSPInstance.SelectButton_Click(Object sender, EventArgs e) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\ChooseKSPInstance.cs:Zeile 74.
   bei System.Windows.Forms.Control.OnClick(EventArgs e)
   bei System.Windows.Forms.Button.OnClick(EventArgs e)
   bei System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   bei System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   bei System.Windows.Forms.Control.WndProc(Message& m)
   bei System.Windows.Forms.ButtonBase.WndProc(Message& m)
   bei System.Windows.Forms.Button.WndProc(Message& m)
   bei System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   bei System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   bei System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   bei System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   bei System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   bei System.Windows.Forms.Form.ShowDialog(IWin32Window owner)
   bei CKAN.Main..ctor(String[] cmdlineArgs, GUIUser User, Boolean showConsole) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Main.cs:Zeile 176.
   bei CKAN.GUI.Main_(String[] args, Boolean showConsole) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Program.cs:Zeile 35.
   bei CKAN.GUI.Main(String[] args) in C:\Users\Martin\Documents\Visual Studio 2015\Projects\CKAN\GUI\Program.cs:Zeile 17.
   bei System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   bei System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   bei Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   bei System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   bei System.Threading.ThreadHelper.ThreadStart() 
```

**Comment by Dazpoet** [comment]
This still requires a rebase and, I assume, a solution to what @Postremus posted above.

**Comment by politas** [comment]
Closing this in favour of #1828, which continues the work
